torch_xla/distributed/xla_backend.py

```diff
@@ -0,0 +1,289 @@
+      input_tensor = torch.cat(input_tensors)
+      reduce_type = self._get_reduce_type(opts.reduceOp)
+      groups = self._mesh
+      shard_count = len(groups[0]) if groups else self.size()
+      result = xm.reduce_scatter(
+          reduce_type,
+          input_tensor,
+          scatter_dim=0,
+          shard_count=shard_count,
+          scale=1,
+          groups=groups)
+
+      output_tensor.set_(result)
+
+    return WorkXla(output_tensors)
+
+  # Call site:
+  # https://github.com/pytorch/pytorch/blob/70f57bcb1e45d21532bdb1c44d3aab018d1cbe88/torch/distributed/distributed_c10d.py#L2683
+  def barrier(self, opts):
+    return WorkXla()
+
+  # Call site:
+  # https://github.com/pytorch/pytorch/blob/70f57bcb1e45d21532bdb1c44d3aab018d1cbe88/torch/distributed/distributed_c10d.py#L1417
+  # `reduce` is not needed by DeepSpeed for now.
+  def reduce(self, *args):
+    raise NotImplementedError
+
+  def allgather_coalesced(self, *args):
+    raise NotImplementedError
+
+  def allreduce_coalesced(self, *args):
+    raise NotImplementedError
+
+  def alltoall(self, *args):
+    raise NotImplementedError
+
+  def alltoall_base(self, *args):
+    raise NotImplementedError
+
+  def gather(self, *args):
+    raise NotImplementedError
+
+  def scatter(self, *args):
+    raise NotImplementedError
+
+  def send(self, *args):
+    raise NotImplementedError
+
+  def recv(self, *args):
+    raise NotImplementedError
+
+  def recv_anysource(self, *args):
+    raise NotImplementedError
+
+  def monitored_barrier(self, *args):
+    raise NotImplementedError
+
+  def Options(self, *args):
+    raise NotImplementedError
+
+
+class WorkXla(Work):
+
+  def __init__(self, cc_tensors=None):
+    '''
+        Args:
+            cc_tensors (List[torch.Tensor]): List of `torch.Tensor`s that
+            have collective communication ops pending.
+            For each Tensor `t` in the list, `t.device` must be an `xla`
+            device.
+        '''
+    super().__init__()
+    self.cc_tensors = cc_tensors
+
+  def wait(self):
+    if self.cc_tensors is not None:
+      if distutils.util.strtobool(
+          os.environ.get('XLA_BACKEND_BLOCKING_CC_OPS', 'False')):
+        logging.info("XLA Backend: Waiting for tensor CC op...")
+        torch_xla._XLAC._xla_sync_multi(self.cc_tensors, devices=[], wait=True)
+      else:
+        logging.info("XLA Backend: Skipping tensor CC op wait.")
+    else:
+      if distutils.util.strtobool(
+          os.environ.get('XLA_BACKEND_BLOCKING_BARRIER', 'False')):
+        logging.info("XLA Backend: Non-tensor wait...")
+        torch_xla._XLAC._xla_step_marker(
+            torch_xla._XLAC._xla_get_default_device(), devices=[], wait=True)
+      else:
+        logging.info("XLA Backend: Skipping non-tensor wait.")
+
+
+# -------------------------------------
+# Override torch.distributed.new_group.
+# -------------------------------------
+_orig_new_group_fn = dist.new_group
+
+
+def infer_mesh(slice_ranks, world_size):
```

Comment:
would I be correct that this is not a user-facing API? If so, we can make it private by adding a "_" prefix?

Reply:
if this is indeed a user-facing API, we might need to 1) add an argument section and a return section; 2) verify that docstrings below will render correctly.

Reply:
No this is not a user-facing API.